From 16480deae0a83e0653f29cc496188c4a2f9569e0 Mon Sep 17 00:00:00 2001
From: Christoph Paasch <cpaasch@openai.com>
Date: Wed, 6 Aug 2025 12:36:16 -0700
Subject: [PATCH 09/17] [ip6gre/ip6vti]: Fix printing of the link-name

When _ATTR_LINK is set, the link needs to be looked up before printing.
Similar as is done for ip6tnl.

So, we need to fix this in ip6gre and ip6vti.

TODO - it looks like we are leaking parent here.... I guess I need to
nl_auto free it... to be confirmed!!!

Signed-off-by: Christoph Paasch <cpaasch@openai.com>
Upstream-Status: Backport [https://github.com/thom311/libnl/commit/e71bdfb1112a04ad48df0a2c3322c4d6cd61e6c7]
Signed-off-by: Jonas Gorski <jonas.gorski@bisdn.de>
---
 lib/route/link/ip6gre.c | 9 ++++++++-
 lib/route/link/ip6tnl.c | 3 ++-
 lib/route/link/ip6vti.c | 9 ++++++++-
 3 files changed, 18 insertions(+), 3 deletions(-)

diff --git a/lib/route/link/ip6gre.c b/lib/route/link/ip6gre.c
index 51d4609cd6bc..aeec1c46d2ea 100644
--- a/lib/route/link/ip6gre.c
+++ b/lib/route/link/ip6gre.c
@@ -246,8 +246,15 @@ static void ip6gre_dump_details(struct rtnl_link *link, struct nl_dump_params *p
 	char addr[INET6_ADDRSTRLEN];
 
 	if (ip6gre->ip6gre_mask & IP6GRE_ATTR_LINK) {
+		struct rtnl_link *parent;
+
 		nl_dump(p, "      link ");
-		name = rtnl_link_get_name(link);
+
+		name = NULL;
+		parent = link_lookup(link->ce_cache, ip6gre->link);
+		if (parent)
+			name = rtnl_link_get_name(parent);
+
 		if (name)
 			nl_dump_line(p, "%s\n", name);
 		else
diff --git a/lib/route/link/ip6tnl.c b/lib/route/link/ip6tnl.c
index 9b3abfd1aaf5..7363753fe7c8 100644
--- a/lib/route/link/ip6tnl.c
+++ b/lib/route/link/ip6tnl.c
@@ -226,9 +226,10 @@ static void ip6_tnl_dump_details(struct rtnl_link *link, struct nl_dump_params *
 {
 	struct ip6_tnl_info *ip6_tnl = link->l_info;
 	char *name, addr[INET6_ADDRSTRLEN];
-	struct rtnl_link *parent;
 
 	if (ip6_tnl->ip6_tnl_mask & IP6_TNL_ATTR_LINK) {
+		struct rtnl_link *parent;
+
 		nl_dump(p, "      link ");
 
 		name = NULL;
diff --git a/lib/route/link/ip6vti.c b/lib/route/link/ip6vti.c
index b6c7a1b2fefe..a6a533191b11 100644
--- a/lib/route/link/ip6vti.c
+++ b/lib/route/link/ip6vti.c
@@ -180,8 +180,15 @@ static void ip6vti_dump_details(struct rtnl_link *link, struct nl_dump_params *p
 	char addr[INET6_ADDRSTRLEN];
 
 	if (ip6vti->ip6vti_mask & IP6VTI_ATTR_LINK) {
+		struct rtnl_link *parent;
+
 		nl_dump(p, "      link ");
-		name = rtnl_link_get_name(link);
+
+		name = NULL;
+		parent = link_lookup(link->ce_cache, ip6vti->link);
+		if (parent)
+			name = rtnl_link_get_name(parent);
+
 		if (name)
 			nl_dump_line(p, "%s\n", name);
 		else
-- 
2.51.0

